### DISCUSSION TITLE
Client side caching eviction policy

### DISCUSSION THREAD

**Discussion by arindamnayak** [redis/rueidis#455] [Q&A] [answered]
I am looking forward to use this library in one of the product. This library provides server assisted client side caching which means it can cache the look up in local memory. I am more interested how the cache eviction happens at local memory. I know there is a setting to control the local cache size ( default value is 128 MB). So when the memory limit is reached, how the data is removed from the local cache? Is there any policy can be set like LRU or LFU?

**Comment by rueian** [comment]
Hi @arindamnayak, It is LRU.

**Reply by arindamnayak** [reply]
It will be helpful if you can point me to the code, where we evict those on condition.

**Reply by rueian** [reply, marked as answer]
Sure!

https://github.com/redis/rueidis/blob/625e68ad6546bfe773ceadc6d3559f3eebdf8b0f/lru.go#L242-L253

**Reply by rueian** [reply]
Hi @arindamnayak, did that help you? I would like to hear about what problem you are facing.

**Reply by arindamnayak** [reply]
Yes, that helped, no problem faced as of now, just exploring the strategy before using it.

**Reply by nikhillimaje** [reply]
Hi @rueian and @arindamnayak, I'm interested in controlling the local cache size for my application. Could you please share how do we do that? Thanks.

**Reply by rueian** [reply]
Hi @nikhillimaje, we have the `CacheSizeEachConn` option for you to control the estimated cache size for each connection.

**Reply by nikhillimaje** [reply]
I see there is an option of specifying `CacheSizeEachConn` which I believe is the cache size for each connection? Instead I'd like to specify the total cache size, like 1GB, irrespective of the number of connections created. Is it possible to do that?

**Reply by rueian** [reply]
Not possible right now. However, the number of connections can be calculated in advance:

Number of connections involved in client-side caching = the number of master redis nodes * 2^PipelineMultiplex.

The default value of PipelineMultiplex in cluster mode is 0, therefore if you have 3 redis masters and want the total cache size to be about 1GB then you can set CacheSizeEachConn to ~340MB.

**Reply by nikhillimaje** [reply]
Got it, thanks for the calculation!

Looking at the system metrics, there are 181 connections created on the application side i.e. the total cache size would be ~22GiB. Previously, we were using another cache (freecache) of size 7GiB to store the redis responses which greatly helped in reducing the qps to Redis. However, after enabling the rueidis client side caching (& disabling freecache), I noticed that the qps to Redis has increased significantly despite the cache size being bigger than the previous one. Any idea why would that be the case?

Also, the way I'm using rueidis client side caching is by simply switching the calls from `DoMulti()` to `DoMultiCache()`. Is there any other config I'm missing?

**Reply by rueian** [reply]
181 connections are quite many. Other settings can affect overall memory impact, for example, the read and write buffer for each connection and the ring queue size of each connection.
https://github.com/redis/rueidis/blob/06ed4fa8cec6fe73754fdb14a9fe0d9840d5d00c/rueidis.go#L121-L134

The size of each cached entry could also be underestimated since we are not able to allocate memory by ourselves, but you can review how we estimate the entry size here https://github.com/redis/rueidis/blob/06ed4fa8cec6fe73754fdb14a9fe0d9840d5d00c/lru.go#L235

**Reply by nikhillimaje** [reply]
I'll look into it. What I'm unable to understand is why I'm not seeing the drop in qps to the Redis with rueidis client side caching.

**Reply by rueian** [reply]
If you have metrics exported for Prometheus or to somewhere, you can record client side hit rate by checking `resp.IsCacheHit()`.

**Reply by rueian** [reply]
> What I'm unable to understand is why I'm not seeing the drop in qps to the Redis with rueidis client side caching.

Do your keys get changed very frequently? If so then that is expected.

**Reply by nikhillimaje** [reply]
Nope. Keys are not changed frequently (only once in 24 hours).

**Reply by rueian** [reply]
Hi @nikhillimaje do you have more findings？

**Reply by nikhillimaje** [reply]
Hi @rueian, I noticed that the CPU util starts to shoot up after a while, irrespective of the cache size I set. This is causing the latency to shoot up eventually. Trying to understand what exactly is increasing the cpu util.

**Reply by rueian** [reply]
It sounds like you have a very low client-side cache hit ratio. It could be caused by:

1. not enough capacity on the client-side cache.
2. short client-side ttl.
3. short server-side ttl.
4. frequent invalidations.

I think we should first check whether you think your client-side cache hit ratio is expected.

For the 4., you can verify it by logging invalidations with the `OnInvalidations` callback.

**Reply by nikhillimaje** [reply]
I added the metric for cache hit ratio. It stayed around 70% for 128MiB as well as 40MiB cache size.

To give you a background, we were using freecache to cache redis result for this particular redis cluster. The size of the freecache was 7GiB and we are seeing a cache hit ratio of 82%. And now I'm trying to replace the freecache with the rueidis client side caching. Now I'm testing again by disabling the caching in rueidis and I no longer see the cpu util spiking up like before (before it went up to 100%).

I'm wondering if I'm misconfiguring something in rueidis client.

Update: The redis cluster we're experimenting with has 90 masters & replicas. Therefore, the rueidis client is creating 180 clients per pod (as we're reading from both master & replica). With 128MiB cache size, the total cache size was around 22GiB. With 40MiB, it was around 7GiB.

**Reply by rueian** [reply]
I guess the cpu spike might be caused by invalidations, could you add two other metrics in `OnInvalidations` like this?

```go
OnInvalidations: func(messages []rueidis.RedisMessage) {
	if messages == nil {
		metrics_csc_full_invalidations.Add(1)	
	} else {
		metrics_csc_invalidations.Add(len(messages))
	}
},
```

**Reply by nikhillimaje** [reply]
The full_invalidations are ~100/min and invalidations are ~1k/sec = ~60k/min which seems to be quite a significant number. What can be done in this case? Is there a way to disable receiving the invalidation msgs or do a noop?

**Reply by nikhillimaje** [reply]
I see there is `ClientTrackingOptions` which is set to OPTIN by default. Will try setting it to an empty array and see if it helps.

**Reply by rueian** [reply]
I believe empty `ClientTrackingOptions` will not help. Actually `OPTIN` is the option that makes you receive the least invalidations.

> invalidations are ~1k/sec = ~60k/min which seems to be quite a significant number. What can be done in this case? Is there a way to disable receiving the invalidation msgs or do a noop?

It indeed seems quite significant, but that is how client-side caching works: whenever a tracked key gets changed, you will receive an invalidation notification to clear the client-side cache. So probably we should revisit two things:

1. What keys can benefit from tracking(client-side caching)? They should be written infrequently compared to reads.
2. Is your server configuration `tracking-table-max-keys` too small? Tracking table overflow also results in pushing invalidation notifications to a client.

**Reply by nikhillimaje** [reply]
Yeah empty `ClientTrackingOptions` didn't help. Our use case is to not track any keys (that's how the current freecache implementation works where keys are evicted based on the lru). Would it be possible to add an option to disable client tracking? I see that rueidis by default sets it to CLIENT TRACKING ON.

**Reply by nikhillimaje** [reply]
I was able to find a workaround by setting the client tracking option to broadcast with some hypothetical prefix. Now the invalidations have dropped to zero and the cpu util is not shooting up. Will monitor for some more time.

**Reply by nikhillimaje** [reply]
Unfortunately, the cpu utilisation still continues to increase even if the invalidation msgs are not received.

I'm curious to know if the cache created is global i.e. if it is shared across all connections or it is used only by the connection that created it?

**Reply by rueian** [reply]
Every connection has its own cache.

> Unfortunately, the cpu utilisation still continues to increase even if the invalidation msgs are not received.

I think that could be caused by the searching through the LRU linked list for item to remove to maintain the cache size:
https://github.com/redis/rueidis/blob/c256ef20a0c980d3a01b33ab845551895bfd49f5/lru.go#L241-L251

Could you can take a CPU profile to identify it?

**Reply by nikhillimaje** [reply]
I couldn't pinpoint if the above code is causing contention, but looking at the profile and top's output, the `runtime.findObject` and `runtime.scanobject` seems to be the heaviest one (which could be related to the above).

```
(pprof) top
Showing nodes accounting for 423.46s, 81.09% of 522.19s total
Dropped 1382 nodes (cum <= 2.61s)
Showing top 10 nodes out of 136
      flat  flat%   sum%        cum   cum%
   193.68s 37.09% 37.09%    200.47s 38.39%  runtime.findObject
    69.70s 13.35% 50.44%     69.85s 13.38%  runtime.(*gcBits).bitp (inline)
    41.51s  7.95% 58.39%     41.68s  7.98%  runtime.(*mspan).heapBitsSmallForAddr
    40.68s  7.79% 66.18%    403.53s 77.28%  runtime.scanobject
    27.27s  5.22% 71.40%     27.27s  5.22%  runtime.(*gcWork).putFast (inline)
    16.51s  3.16% 74.56%     16.51s  3.16%  runtime.(*mspan).base (inline)
    14.25s  2.73% 77.29%     14.25s  2.73%  runtime/internal/atomic.(*Uint32).CompareAndSwap (inline)
     6.95s  1.33% 78.62%      6.95s  1.33%  runtime/internal/syscall.Syscall6
     6.94s  1.33% 79.95%      6.94s  1.33%  runtime/internal/atomic.(*Uint64).CompareAndSwap
     5.97s  1.14% 81.09%     13.16s  2.52%  runtime.(*activeSweep).end
```

<img width="260" alt="Screenshot 2024-11-29 at 15 11 39" src="https://github.com/user-attachments/assets/8af19586-b499-4f77-961a-26b14ac892d8">

**Reply by rueian** [reply]
okay, that suggests the built-in LRU put a lot of pressure on GC. I am afraid this can not be improved in a short time but you can replace it with freecache by implementing the `SimpleCache` interface and put the implementation with the `NewCacheStoreFn` and `NewSimpleCacheAdapter`:

```go
type SimpleCache interface {
	Get(key string) RedisMessage
	Set(key string, val RedisMessage)
	Del(key string)
	Flush()
}

client, err := rueidis.NewClient(rueidis.ClientOption{
	NewCacheStoreFn: func(option rueidis.CacheStoreOption) rueidis.CacheStore {
		return rueidis.NewSimpleCacheAdapter(YOUR_IMPL_HERE)
	},
})

**Reply by nikhillimaje** [reply]
Tried the above. One of the difficulties I'm facing is to convert RedisMessage to byte[] and back, as the freecache only accepts ([]byte, []byte). I'm only interested in the string part of RedisMessage for which I can do val.AsBytes() in `Set`. But unable to set this byte array as a string in RedisMessage in `Get`.

**Reply by rueian** [reply]
Hi @nikhillimaje, we are going to have that ability in the next release https://github.com/redis/rueidis/pull/693.

**Reply by nikhillimaje** [reply]
Thanks @rueian!
